### DISCUSSION TITLE
Setting up local printers on the clients

### DISCUSSION THREAD

**Discussion by samthelion17** [ltsp/ltsp#158] [Q&A]
Hi, I've been trying to set up local printers on LTSP clients (using Ubuntu 20.04.1, with latest ltsp from ppa, chrootless environment). The Issue #8 was really helpful helping me understand how the new LTSP works but I'm , clearly, missing something. Can somebody point me out in the right direction?

Best regards!

**Comment by alkisg** [comment]
You're supposed to put CUPS_SERVER=localhost in ltsp.conf for that. Then LTSP clients behave like standalone clients, i.e. they can share their local printers or see other shared printers from the network.

**Comment by samthelion17** [comment]
> You're supposed to put CUPS_SERVER=localhost in ltsp.conf for that. Then LTSP clients behave like standalone clients, i.e. they can share their local printers or see other shared printers from the network.

But then I'll have to set up the printer(s) every time after client boot-up, right?

**Comment by alkisg** [comment]
If the image has the drivers, the printers should be automatically activated when cups starts.
If you want to forward the printer ports to the server, like LTSP5 did with jetpipe, you can either manually download jetpipe.py, or install and use [p910nd](https://manpages.debian.org/p910nd) instead of jetpipe.

**Comment by samthelion17** [comment]
Ok, I will try that soon but, just in case I need to use p910nd, let's sketch out the procedure of using p910nd with the new LTSP. I remember that back in the days of LTSP5 printers were added in the lts.conf, but things have changed, so the right steps to do now would be:

1. Install p910nd
2. Update image
3. ?

**Comment by alkisg** [comment]
3. If p910nd has a systemd service etc, it will just need to be enabled.
If it's a standalone program like jetpipe.py, you'd run it via ltsp.conf like
POST_SERVICE_P910ND="p910nd parms &"

**Comment by samthelion17** [comment]
4. Go to CUPS admin / Add Printer
5. Select 'AppSocket/HP JetDirect'
6. Fill connection with the right data "socket://<client_IP>:<port>", port being 910{0..9}
7. Setup driver and other settings...

But still, I need to tell p910nd which printer corresponds to a specific port. Let's assume that all my printers are USB printers and that every client will have only one printer connected to it.

**Comment by alkisg** [comment]
This is the "params" in the p910nd command line; you can specify the printer path from there, e.g. /dev/usb/lp0. Same with jetpipe.py.

**Comment by alkisg** [comment]
Discussions are now enabled for the LTSP repository.
I converted this issue into a discussion.

**Comment by wfragakis** [comment]
fwiw, just did an install on Debian 10. Not sure if I installed too much but it works (generally).
1. Installed p910nd via apt
2. Installed cups (not sure if required)
3. Edited /etc/default/p910nd based on docs in /usr/share/docs/p910nd
```

# Printer number, if not 0
P910ND_NUM="0"
# Additional daemon arguments, see man 8 p910nd
P910ND_OPTS=" -f /dev/usb/lp1"

# Debian specific (set to 1 to enable start by default)
P910ND_START=1

```
Added CUPS_SERVER=localhost to /etc/ltsp/ltsp.conf as advised above

We use a separate server as the main print server which has no problems with creating and sending print jobs to the clients.

One problem I encountered is that if one uses wireless keyboards, the printers invariably end up as /dev/usb/lp1 instead of lp0 with wired keyboards. Not sure how to set config for using one or the other while using this config which is easy because p910nd runs as a service without too much intervention. I presume I could use the POST_SERVICE_P910ND="p910nd parms &" but, for our setup (medical facility), we are mostly wireless keyboards so I didn't pursue this.
